src/backend/distributed/utils/maintenanced.c

```diff
@@ -292,45 +329,74 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 	BackgroundWorkerHandle *backgroundTasksQueueBgwHandle = NULL;
 	bool backgroundTasksQueueWarnedForLock = false;
 
-	/*
-	 * We do metadata sync in a separate background worker. We need its
-	 * handle to be able to check its status.
-	 */
-	BackgroundWorkerHandle *metadataSyncBgwHandle = NULL;
 
-	/*
-	 * Look up this worker's configuration.
-	 */
-	LWLockAcquire(&MaintenanceDaemonControl->lock, LW_EXCLUSIVE);
+	MaintenanceDaemonDBData *myDbData = NULL;
 
-	MaintenanceDaemonDBData *myDbData = (MaintenanceDaemonDBData *)
-										hash_search(MaintenanceDaemonDBHash, &databaseOid,
-													HASH_FIND, NULL);
-	if (!myDbData)
+	if (databaseOid == 0)
 	{
-		/*
-		 * When the database crashes, background workers are restarted, but
-		 * the state in shared memory is lost. In that case, we exit and
-		 * wait for a session to call InitializeMaintenanceDaemonBackend
-		 * to properly add it to the hash.
-		 */
+		/* TODO : Get the admin database name from GUC contro_db*/
+		char *databaseName = "postgres";
 
-		proc_exit(0);
-	}
+		BackgroundWorkerInitializeConnection(databaseName, NULL, 0);
+
+		/* Now we have a valid MyDatabaseId. */
+		/* Insert the daemon instance to the hash table. */
+		bool found;
+
+		LWLockAcquire(&MaintenanceDaemonControl->lock, LW_EXCLUSIVE);
+
+		myDbData = (MaintenanceDaemonDBData *) hash_search(
+			MaintenanceDaemonDBHash,
+			&MyDatabaseId,
+			HASH_ENTER_NULL,
+			&found);
 
-	if (myDbData->workerPid != 0)
+		if (!myDbData)
+		{
+			LWLockRelease(&MaintenanceDaemonControl->lock);
+			return;
+		}
+
+		if (!found)
+		{
+			/* ensure the values in MaintenanceDaemonDBData are zero */
+			memset(((char *) myDbData) + sizeof(Oid), 0,
+				   sizeof(MaintenanceDaemonDBData) - sizeof(Oid));
+		}
+
+		before_shmem_exit(MaintenanceDaemonShmemExit, ObjectIdGetDatum(MyDatabaseId));
+		databaseOid = MyDatabaseId;
+		myDbData->userOid = GetSessionUserId();
+	}
+	else
 	{
-		/*
-		 * Another maintenance daemon is running. This usually happens because
```

Comment:
This handling of an edgecase seems to have been removed by accident. And this same edgecase should be handled for the control database.

Reply:
I had merged those two error cases. But yes we have lost the comments section. So I restored.

In If (databaseOid = 0) block workerPid zeroed out by memset. So I am not adding if it is non-zero here check.